### DISCUSSION TITLE
make samples/sensor/ds18b20 work with ESP32

### DISCUSSION THREAD

**Discussion by n1ghtrunner** [zephyrproject-rtos/zephyr#51987] [Q&A]
## Problem:
- Try to read d18b20 with the wiring according to sample description.
- The wiring works for nrf52840 soc. Tested with nrf52840dk (via provided overlay) and with nrf52840_mdk_usb_dongle (with a custom overlay file)
- I cannot make it work for the ESP32 NodeMCU board.
- Problems seems to be that either the ESP32 isn't able to handle it from the electrical perspective or I didn't define the overlay right.
- Of course tried different overlay configurations, partially modifying the properties of uart2/group1 and uart2/group2, but no success in any combination.

## Scenario Description:
- Sample: ```sensors/ds18b20```
- Board: AI-Thinker ESP32 NodeMCU Dev Board. 
  - Pinout looks like [this one](https://lastminuteengineers.com/esp32-pinout-reference/)
- ds18b20 wiring:
  -  one single sensor
  - 1-wire bus line wired to uart2 (on gpio pins 16/17 - both directly connected). 
  - ds18b20 V+ connected to 3.3v V+ of NodeMCU board. GND accordingly.
  - 4.7kΩ between 1-wire bus line and V+
  - Also tried variations with schottky diode (with optional 220Ω/480Ω) between 1-wire bus line and TX (on ESP)
- Using board configuration: ```esp32```
- Using custom overlay file: ```<sample directory>/boards/esp32.overlay```

Content of ```esp32.overlay```:
```
&uart2 {
	status = "okay";

	w1: w1 {
		compatible = "zephyr,w1-serial";
		status = "okay";

		ds18b20 {
			compatible = "maxim,ds18b20";
			family-code = <0x28>;
			resolution = <12>;
			status = "okay";
		};
	};
};

&uart2_default { 
  group1 {
    bias-pull-up;
    drive-open-drain;
  };
};
```
## Alternative solution without zephyr
* There is a 3 year old code to make the same setup work with esp-idf: [esp-idf-onewire](https://github.com/StanislavLakhtin/esp-idf-onewire). I tried to compile this with recent and some older versions of esp-idf to validate the electric setup. However no success - always fails with some linking error at the end of make execution. 
* The author suggests to have 1-wire on 5v (with signal level converter between ESP32 and 1-wire) and not 3.3v. However, since the 3.3.v setup works flawlessly on the nordic boards this does not seem to be the issue.

**Comment by rftafas** [comment]
Hi!

FWIW, alternative ESP-IDF solution would be this one: https://github.com/espressif/esp-idf/tree/master/examples/peripherals/rmt/onewire_ds18b20

**Reply by n1ghtrunner** [reply]
Of course any of the examples that use bit-banging on an arbitrary GPIO port work easily (esp-idf, Arduino, ...). Nevertheless Zephyr supports only 1-Wire on UART. I do not want to solve the problem - I want to solve it with zephyr to achieve a solution where I can reuse my partial solution for different boards (nrf52840, esp32, ...).

**Reply by rftafas** [reply]
If the bit banging example works, at least for a slow signal you will know it is electrically right. Then you can go back to Zephyr and test over the same setup. I would suggest starting with same speed used on bit banging.

If even the bit-banging is not doing it, there might be other issues.

Also, you may already have your Zephyr updated, but always worth suggesting using the latest version.
